appjs/views/edit_project.js

```diff
@@ -58,6 +134,49 @@ var EditProject = BaseView.extend(require('./mixins/actions'), require('./mixins
     }, this);
   },
 
+  askToSave: function() {
+    logger.debug('got to ask to save');
+    var view = this;
+    var modalView = new BaseModalView({app: this.app, parentView: this}),
+        ret = new Promise(function(resolve, reject) {
+          modalView.once('cancel submit', function() {
+            resolve(true);
+          });
+          modalView.on('close', function() {
+            logger.debug('~~~~ got close');
+            resolve(false);
+          });
+        });
+
+    if($('#base-modal').length === 0){
+      modalView.show();
+    }
+
+    return ret;
+  },
+
+  hasUnsavedChanges: function(){
+    if(!this.upToDate){
+      return true;
+    } else {
+      return false;
+    }
+  },
+
+  determineEquality: function(newData){
+    var view = this;
+    // set view.formDataOnLoad in afterRender - form validation (view.model.formData())
+    // converts [0] values to null, which causes the objects to validate as unequal
+    // even though they really are equal
+    if(_.isEqual(view.formDataOnLoad, newData) ){
```

Comment:
The structure of the data when using `this.model.formData()` doesn't always match the structure of the data you get when using `data = $form.alpaca('get').getValue();`. While `$form.alpaca('get').getValue()` returns empty values as empty arrays, `this.model.formData()` returns them as `null` values. I'd originally written a block of code to correct for this, but realized that it quickly became more complicated than it really ought to be since you'd have to search through nested key/val variables until reaching the innermost level.